### ISSUE TITLE
[RPi4] Enabling XInput support in Wine

### ISSUE THREAD

**Issue by DevilBlackDeath** [ptitSeb/box86#809]
Hi,
Sorry for posting here as this is not meant to be a support forum afaik though it could be a bug. I honestly couldn't find information on it.

Box86 is running on a 4GB Raspberry Pi 4 running Raspbian legacy buster (the one available in the RPi Imager) with desktop environment.
I'm using Box86 to run wine 8.3 devel but I can't seem to be able to enable controller support. On my x86 computer the same 8.0 stable reports my controller in 'wine control joy.cpl', but no luck so far on the pi. I've tried running xboxdrv with and without mimicing xpad and detaching kernel drivers respectively. I've tried no driver at all (afaik wine 8 supports xinput by default).

I know about logs but I wouldn't know which are useful in this scenario. I might have missed something about how to enable that but I've read over the readmes multiple times and the website and couldn't find anything about controller or xinput support. I couldn't find issues mentioning it or any of the setup instructions. I assume that may mean it should work by default which is why I'm posting it as an issue in the first place

Of note is that I also tried with wine 8.0 stable, 8.3 staging, 5.5 staging and 7.1 devel.

**Comment by ptitSeb** [comment]
there might be a missing wrapped lib or wrapped function.
Try again uing `BOX86_LOG=1` and `BOX86_DLSYM_ERROR=1` and see if some lib or function is missing from the generated logs.

**Comment by DevilBlackDeath** [comment]
Hey thanks for answering so fast. I ran both commands and that's the log I get.

[xinput_log.txt](https://github.com/ptitSeb/box86/files/11170452/xinput_log.txt)

I haven't used those before so I'm not 100% sure what to look for but I saw nothing that seemed to mention missing symbols or libraries.

**Comment by ptitSeb** [comment]
Well, the logs doesn't show any missing libs or wrapped function.
The wineusb not starting might explain, but U'm not sure how wine handle joystick, so I'll probably have to check and reproduce myself.

Just to be sure: your joystick is recognise by regular linux apps right? Also, it would be interesting to see if the joystick work with non-wine x86 games too.

**Comment by DevilBlackDeath** [comment]
I'll have to get back on you about non-wine x86 games but yeah regular linux apps detect it. I use the Retropie script and lzdoom and emulators detect it fine. I have also run Neurovoider and Celeste using the mono/FNA trick and those detect the controller as well. Do you have any easy to access game in mind to test x86 natives ?

**Comment by ptitSeb** [comment]
Not really (but you can try this one: https://www.lexaloffle.com/neko.php I think joystick is supported). The point is more to be sure that current OS sees the Joystick fine (for example, on my current ARM64 test machine running Debian, the Kernel doesn't have any joystick driver, and I cannot use any of them).

**Comment by DevilBlackDeath** [comment]
I was actually able to find a game myself on itch.io (didn't expect there to be so many linux native games with gamepad support, though it was a bit more troublesome to find one made for x86/i386. EGGNOGG (Classic) has controller support. Ran it on box86 on my Pi 4 and the controller works perfectly.

So it works with FNA/Mono games, native armhf apps like lzdoom and Retroarch/libretro and it even works on box86 natives apparently. Really comes down to a Wine compatibility issue it would seem. As I said I believe Wine is somewhat able to still detect a xinput controller even without kernel drivers, so I'd say chances are it does something weird with it (then again, FNA/Mono games seem to be able to use even when I don't have drivers installed or running, so they might do the same, except they do it natively).

I know that before Wine added proper xinput support there were a few attempt at adding xinput support through SDL2 and some of the sources are still available ! I'll try and see if those could still work with recent Wine versions, and if so, if that could solve the issue. Will report back on that.

**Comment by DevilBlackDeath** [comment]
Ok, so it doesn't seem to properly compile anymore. However the exact project was koku-xinput-wine. Wouldn't it be possible for Box86, since it probably already wraps part of SDL, to use the same techniques to transfer xinput to Wine ?

I'll also try and see if using the controller wireless fixes the issue. Hopefully this would not be handled by wineusb or winebus ? And could further determine the issue.

**Comment by ptitSeb** [comment]
I would prefer for the joystick to just work the way if works on x86

**Comment by DevilBlackDeath** [comment]
Makes sense ! Plus that would probably not be a good viable long-term solution. Well in all cases, if it can work there's a bunch of games that are gonna go on my Pi. Tests with box86 so far are super promising when it comes to any 2D games with a sub 2-2.1Ghz requirement. Most XNA (the non-FNA ported ones, or even the ones that were ported but didn't recompile the FNA version for Windows, there's actually quite a few in my library that do this) games work out of the box with little performance impact. Anything that runs natively using fairly standard libraries should also work pretty well so I'm looking forward to test that now that I've confirmed controller compatibility on that end.

At least we know that Box86 sees the controller alright and it's only a Wine - Box86 issue so far.

**Comment by DevilBlackDeath** [comment]
Further research revealed that using box64+wine64 seems to recognize the input. Didn't test in-game yet, but 'wine control joy.cpl' properly lists the gamepad, and the XInput tab recognized all inputs from the gamepad.

This was tested with wine 64 8.4 devel. Whether it is thanks to the 8.4 update (as of yet unavailable for i386) or thanks to box64 itself I don't know (I'd wager it's probably box64 doing something box86 doesn't do yet or does differently). It's worth noting my standard x86/x64 tests were, contrary to what I said (honest mistake sorry), executed in a X64 environment. As a result it could be that wine x86 is simply incapable of detecting gamepads ? I don't have an x86 linux environment readily available so I can't test that unfortunately.

**Comment by ptitSeb** [comment]
Any news on this ticket or can it be closed?
